source/common/formatter/substitution_formatter.cc

```diff
@@ -54,5 +54,101 @@ const re2::RE2& SubstitutionFormatParser::commandWithArgsRegex() {
   // clang-format on
 }
 
+JsonFormatBuilder::FormatElements
+JsonFormatBuilder::fromStruct(const ProtobufWkt::Struct& struct_format) {
+  output_.clear();
+
+  formatValueToFormatElements(struct_format.fields());
+  std::string json_piece = std::move(buffer_);
+  if (!json_piece.empty()) {
+    output_.push_back(JsonString{std::move(json_piece)});
+  }
+  return std::move(output_);
+};
+
+void JsonFormatBuilder::formatValueToFormatElements(const ProtobufWkt::Value& value) {
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::KIND_NOT_SET:
+  case ProtobufWkt::Value::kNullValue:
+    serializer_.addNull();
+    break;
+  case ProtobufWkt::Value::kNumberValue:
+    // TODO(wbpcode): may use Buffer::Util::serializeDouble for best performance.
+    // But this function only be executed when the configuration is loading. So,
+    // it's fine to use the fmt::to_string.
+    keep_value_type_ ? serializer_.addNumber(value.number_value())
+                     : serializer_.addString(fmt::to_string(value.number_value()));
+    break;
+  case ProtobufWkt::Value::kStringValue: {
+    absl::string_view string_format = value.string_value();
+    if (!absl::StrContains(string_format, '%')) {
+      serializer_.addString(string_format);
+      break;
+    }
+
+    // The string contains a formatter, we need to push the current raw string
+    // into the output list first.
+    std::string json_piece = std::move(buffer_);
+    if (!json_piece.empty()) {
+      output_.push_back(JsonString{std::move(json_piece)});
+    }
+
+    // Now a formatter is coming, we need to push the current raw string into
+    // the output list.
+    output_.push_back(TmplString{std::string(string_format)});
+    break;
+  }
+  case ProtobufWkt::Value::kBoolValue:
+    keep_value_type_ ? serializer_.addBool(value.bool_value())
+                     : serializer_.addString(value.bool_value() ? Json::Constants::True
+                                                                : Json::Constants::False);
+
+    break;
+  case ProtobufWkt::Value::kStructValue: {
+    formatValueToFormatElements(value.struct_value().fields());
+    break;
+  case ProtobufWkt::Value::kListValue:
+    formatValueToFormatElements(value.list_value().values());
+    break;
+  }
+  }
+}
+
+void JsonFormatBuilder::formatValueToFormatElements(const ProtoList& list_value) {
+  buffer_.append(Json::Constants::ArrayBegin); // Delimiter to start list.
```

Comment:
See, we have make the `addWithoutSnaitizing` a private method so we cannot use it? And the reasons why I finally made it a private method are:

1. It's unusual (or even dangerous) (public) method for JSON streamer. Un-careful calling to that method may break the whole output.
2. For cases where we actually need to push raw data to the buffer, the developers self can do it by operating the underlaying buffer directly and take responsibility to the operation.

The way of using the `Streamer` in this PR is a un-common. Because we need low level control to the delimiters to ensure the output array of format string and raw JSON piece is correct.

Reply:
I'd make it public and add the assert per Alyssa's suggestion.

I think everything shoudl go through the streamer API.

Reply:
Why? I think that makes no sense for both the Streamer and this feature?

It cannot make the Stramer better and cannot make this code simpler.

I have tried to use the Streamer as much as possible, but things doesn't become simpler, and I need to add some new methods to Streamer to expose more statuses.

It just make both Streamer and this code weird.

So, I finally just use the Streamer as a helper to sanitize value, and it work very well.

Reply:
I feel pretty strongly that (a) we should delegate to the Serializer all the json serialization logic and (b) we should not undermine the Serializer's control by writing directly to its buffer. That's an abstraction-boundary violation IMO. The serializer should be able to decide to buffer some bounded-size things if it wants to, though it doesn't do this now.

Reply:
To a):

Per my previous comment, I don't think it's a good idea to make the `addWithoutSanitizing` public. An acceptable way to me is create a derived class from Streamer which is used only by the formatter and make the formatter as a friend class of the derived class. Then, we can use the `addWithoutSanitizing` in the formatter but cannot use it in other position.

To b):

See https://github.com/envoyproxy/envoy/pull/35545#discussion_r1752315968. I tried to use Streamer to serialize everything after the template Streamer was landed, but it make things more complex and require Streamer to provide more inner status. So, I back to previous way.

The final target is only one: Create a JSON formatter which have comparable performance with TEXT formatter, based on the `Protobuf::Struct` configuration.

I am pretty happy to see better design or implementation if you could figure out one.

There are lots comment to present the core idea of this PR, see this https://github.com/envoyproxy/envoy/pull/35545#issuecomment-2272759820 or this https://github.com/envoyproxy/envoy/pull/35545#discussion_r1758060206. May could used as a reference for you.